### DISCUSSION TITLE
MySQL (Vitess) Support, removing use of foreign key constraints

### DISCUSSION THREAD

**Discussion by janekwunderlich** [directus/directus#13395] [Feature Requests]
A lot of people are using Vitess nowadays to make MySQL scalable. For several reasons Vitess decided to drop support for foreign key constraints and as a reason, trying to install system tables of Directus on a MySQL database running on Vitess fails.

You can read more on their reasoning here: https://vitess.io/blog/2021-06-15-online-ddl-why-no-fk/#:~:text=FOREIGN%20KEY%20Constraints%20overview%20%23,the%20enforcement%20of%20that%20relationship.

Log output
`
19/05/2022, 03:13:08 | Error: ER_QUERY_INTERRUPTED: foreign key constraints are not allowed, see https://vitess.io/blog/2021-06-15-online-ddl-why-no-fk/ | directus
-- | -- | --
19/05/2022, 03:13:08 | at Query.Sequence._packetToError (/directus/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14) | directus
19/05/2022, 03:13:08 | at Query.ErrorPacket (/directus/node_modules/mysql/lib/protocol/sequences/Query.js:79:18) | directus
19/05/2022, 03:13:08 | at Protocol._parsePacket (/directus/node_modules/mysql/lib/protocol/Protocol.js:291:23) | directus
19/05/2022, 03:13:08 | at Parser._parsePacket (/directus/node_modules/mysql/lib/protocol/Parser.js:433:10) | directus
19/05/2022, 03:13:08 | at Parser.write (/directus/node_modules/mysql/lib/protocol/Parser.js:43:10) | directus
19/05/2022, 03:13:08 | at Protocol.write (/directus/node_modules/mysql/lib/protocol/Protocol.js:38:16) | directus
19/05/2022, 03:13:08 | at TLSSocket.ondata (node:internal/streams/readable:754:22) | directus
19/05/2022, 03:13:08 | at TLSSocket.emit (node:events:527:28) | directus
19/05/2022, 03:13:08 | at addChunk (node:internal/streams/readable:315:12) | directus
19/05/2022, 03:13:08 | at readableAddChunk (node:internal/streams/readable:289:9) | directus
19/05/2022, 03:13:08 | -------------------- | directus
19/05/2022, 03:13:08 | at Protocol._enqueue (/directus/node_modules/mysql/lib/protocol/Protocol.js:144:48) | directus
19/05/2022, 03:13:08 | at Connection.query (/directus/node_modules/mysql/lib/Connection.js:198:25) | directus
19/05/2022, 03:13:08 | at /directus/node_modules/knex/lib/dialects/mysql/index.js:132:18 | directus
19/05/2022, 03:13:08 | at new Promise (<anonymous>) | directus
19/05/2022, 03:13:08 | at Client_MySQL._query (/directus/node_modules/knex/lib/dialects/mysql/index.js:126:12) | directus
19/05/2022, 03:13:08 | at executeQuery (/directus/node_modules/knex/lib/execution/internal/query-executioner.js:37:17) | directus
19/05/2022, 03:13:08 | at Client_MySQL.query (/directus/node_modules/knex/lib/client.js:146:12) | directus
19/05/2022, 03:13:08 | at Runner.query (/directus/node_modules/knex/lib/execution/runner.js:130:36) | directus
19/05/2022, 03:13:08 | at Runner.queryArray (/directus/node_modules/knex/lib/execution/runner.js:224:21) | directus
19/05/2022, 03:13:08 | at Runner.queryArray (/directus/node_modules/knex/lib/execution/runner.js:276:31) { | directus
19/05/2022, 03:13:08 | code: 'ER_QUERY_INTERRUPTED', | directus
19/05/2022, 03:13:08 | errno: 1317, | directus
19/05/2022, 03:13:08 | sqlMessage: 'foreign key constraints are not allowed, see https://vitess.io/blog/2021-06-15-online-ddl-why-no-fk/', | directus
19/05/2022, 03:13:08 | sqlState: '70100', | directus
19/05/2022, 03:13:08 | index: 0, | directus
19/05/2022, 03:13:08 | sql: 'alter table `directus_users` add constraint `directus_users_role_foreign` foreign key (`role`) references `directus_roles` (`id`)' | directus
19/05/2022, 03:13:08 | } | directus
19/05/2022, 03:13:08 | 01:13:08 ✨ Installing Directus system tables... | directus

`

**Comment by rijkvanzanten** [comment]
This is also preventing running Directus on PlanetScale (same reason)

**Reply by janekwunderlich** [reply]
Do you think opening a discussion on knex would make sense? I see that they resolved/created the custom CockroachDB Client because of an issue people ran into. @rijkvanzanten

**Reply by rijkvanzanten** [reply]
I don't think this is a problem to solve on Knex's side of things 🤔 Knex doesn't force anybody to use foreign keys, so to them this issue is as simple as "just don't use .foreign()". 

The solution here would be to effectively reimplement all foreign key constraint magic on the Directus API side, and then add an environment variable to allow enabling/disabling syncing of the actual foreign key constraints (and triggers) to the database. No small feat!

**Reply by janekwunderlich** [reply]
I mean foreign keys still work, just not constraints on foreign keys as they cause significant performance issues on write operations on big databases/if the constraint affects a lot of related fields. What they suggest is running those deletions or whatever it is that the foreign key constraint would cause on off-peak times through own logic (as you described). 

The solution is definitely not that easy and would require quite some additional logic, it would still be great to see compatibility sometime in the future though! Especially as PlanetScale is gaining big in popularity right now.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by claudio-silva** [comment]
Planetscale now has experimental support for Foreign Key Constraints.
https://planetscale.com/docs/concepts/foreign-key-constraints
Unfortunately, even with that enabled, Directus still fails on the initial setup of the database with a different error not related to foreign keys.
